python-api/livy/client.py

```diff
@@ -0,0 +1,444 @@
+
+    def __init__(self, uri, **kwargs):
+        self._config = SafeConfigParser()
+        self._load_config(**kwargs)
+        match = re.match(r'(.*)/sessions/([0-9]+)', uri.path)
+        if match:
+            base = ParseResult(scheme=uri.scheme, netloc=uri.netloc,
+                               path=match.group(1), params=uri.params,
+                               query=uri.query, fragment=uri.fragment)
+            self._set_uri(base)
+            self._conn = LivyConnection(base)
+            self._session_id = int(match.group(2))
+            self._reconnect_to_existing_session()
+        else:
+            self._set_uri(uri)
+            session_conf_dict = dict(self._config.items(self._CONFIG_SECTION))
+            self._conn = LivyConnection(uri)
+            self._session_id = self._create_new_session(
+                session_conf_dict).json()['id']
+        self._executor = ThreadPoolExecutor(max_workers=1)
+        self._stopped = False
+
+    def submit(self, job):
+        """
+        Submits a job for execution to the spark cluster.
+
+        Parameters
+        ----------
+        job : function
+            The function must accept a single parameter, which is an instance
+            of JobContext. Spark jobs can be created with the help of
+            JobContext, which exposes the Spark libraries
+
+        Returns
+        -------
+        job_handle : an instance of the class JobHandle
+            A handle that can be used to monitor the job
+
+        Examples
+        -------
+        >>> def simple_spark_job(context):
+        >>>     elements = [10, 20, 30, 40, 50]
+        >>>     return context.sc.parallelize(elements, 2).count()
+
+        >>> client.submit(simple_spark_job)
+
+        """
+        return self._send_job('submit-job', job)
+
+    def run(self, job):
+        """
+        Asks the remote context to run a job immediately.
+
+        Normally, the remote context will queue jobs and execute them based on
+        how many worker threads have been configured. This method will run
+        the submitted job in the same thread processing the RPC message,
+        so that queueing does not apply.
+
+        It's recommended that this method only be used to run code that
+        finishes quickly. This avoids interfering with the normal operation
+        of the context.
+
+        Parameters
+        ----------
+        job : function
+            The function must accept a single parameter, which is an instance
+            of JobContext. Spark jobs can be created with the help of
+            JobContext, which exposes the Spark libraries.
+
+        Returns
+        -------
+        future : concurrent.futures.Future
+            A future to monitor the status of the job
+
+        Examples
+        -------
+        >>> def simple_spark_job(context):
+        >>>     elements = [10, 20, 30, 40, 50]
+        >>>     return context.sc.parallelize(elements, 2).count()
```

Comment:
It would be nice to use an example that actually follows the "should finish quickly" recommendation set by the documentation.